Data/ByteString/Lazy/ReadInt.hs

```diff
@@ -0,0 +1,229 @@
+    -- of leading zeros on a 32-bit system!  Chunk lengths are
+    -- Ints, while lazy bytestring lengths are Int64s.  With
+    -- @'(.|.)'@ we get to /unask/ any questions about overflow.
+    --
+  where
+    consume :: ByteString -> Int -> Accum -> Maybe (a, ByteString)
+    -- All done
+    consume Empty !valid !acc
+        = if valid /= 0 then convert acc Empty else Nothing
+    -- Process next chunk
+    consume (Chunk c@(BI.BS _ !len) cs) !valid !acc
+        = case _digits q r c acc of
+            Result used acc'
+                | used == len
+                  -- process remaining chunks
+                  -> consume cs (valid .|. used) acc'
+                | (valid .|. used) /= 0
+                  -- ran into a non-digit
+                  -> convert acc' $
+                     let !c' = BU.unsafeDrop used c in Chunk c' cs
+            _     -> Nothing
+#endif
+    convert :: Accum -> ByteString -> Maybe (a, ByteString)
+    convert !acc !s =
+        let !i = case r of
+                -- See [Note @maxBound `quotRem` 10 == (q, 7)@]
+                7 -> fromIntegral @Accum @a acc
+                _ -> negate @a $ fromIntegral @Accum @a acc
+         in Just (i, s)
+
+----- Per chunk decoder
+
+data Result = Overflow | Result !Int !Accum
+
+-- | Process as many digits as we can, returning the additional
+-- number of digits found and the updated accumulator.  If the
+-- accumulator would overflow return 'Overflow'.
+--
+_digits :: Accum -> Accum -> BI.ByteString -> Accum -> Result
+{-# INLINE _digits #-}
+_digits !q !r !(BI.BS !fp !len) = \ !acc ->
+    BI.accursedUnutterablePerformIO $
+        BI.unsafeWithForeignPtr fp $ \ptr -> do
+            let end = ptr `plusPtr` len
+            go ptr end ptr acc
+  where
+    go start end = loop
+      where
+        loop !ptr !acc = getDigit >>= \ !d ->
+            if | d > 9
+                 -> return $ Result (ptr `minusPtr` start) acc
+               | acc < q || acc == q && d <= r
+                 -> loop (ptr `plusPtr` 1) (acc * 10 + d)
+               | otherwise
+                 -> return Overflow
+          where
+            fromDigit = \w -> fromIntegral w - 0x30 -- i.e. w - '0'
+            --
+            getDigit :: IO Accum
+            getDigit | ptr /= end = fromDigit <$> peek ptr
+                     | otherwise  = pure 10  -- End of input
+            {-# NOINLINE getDigit #-}
+            -- The getDigit function makes it possible to implement a single
+            -- success exit point from this function.  If we return 'Result'
+            -- from multiple places in the code, when `_digits`, is inlined
+            -- it generates (at least GHC 8.10 through 9.2) separate copies
+            -- of the continuation code that processes the return value.
+            --
+            -- The NOINLINE pragma is required to avoid inlining branches
+            -- that would restore multiple exit points.
```

Comment:
The work-around is not brittle, because this is not different from `consume` being inlined into `_digits` as a bunch of join points.

But for the record, changing the `Result -> ...` continuation to a NOINLINE function, and having to exit points does reduce code duplication to a degree, because the resulting duplicate join points no longer copy the continuation, they branch to it instead, but I'm inclined to leave the code as-is.

See however: https://gitlab.haskell.org/ghc/ghc/-/issues/20717